diff -Nur kernel-2.6.28.10-vs2.3.0.36.11.85b15/arch/ia64/kernel/mca_drv.c kernel-2.6.28.10-vs2.3.0.36.11.85b15_fmd/arch/ia64/kernel/mca_drv.c
--- kernel-2.6.28.10-vs2.3.0.36.11.85b15/arch/ia64/kernel/mca_drv.c	2011-03-29 18:04:19.000000000 +0800
+++ kernel-2.6.28.10-vs2.3.0.36.11.85b15_fmd/arch/ia64/kernel/mca_drv.c	2011-06-30 11:43:45.000000000 +0800
@@ -503,12 +503,10 @@
 	if (!evt)
 		return;
 
-	evt->cpu = cpuid;
-	evt->fullclass = fullclass;
+	evt->cpuid = cpuid;
+	strcpy(evt->eclass, fullclass);
 
 	fm_event_post_intr(evt);
-
-	kfree(evt->fullclass);
 	kfree(evt);
 }
 
diff -Nur kernel-2.6.28.10-vs2.3.0.36.11.85b15/drivers/ata/libata-eh.c kernel-2.6.28.10-vs2.3.0.36.11.85b15_fmd/drivers/ata/libata-eh.c
--- kernel-2.6.28.10-vs2.3.0.36.11.85b15/drivers/ata/libata-eh.c	2011-03-29 18:04:19.000000000 +0800
+++ kernel-2.6.28.10-vs2.3.0.36.11.85b15_fmd/drivers/ata/libata-eh.c	2011-06-30 11:35:28.000000000 +0800
@@ -1011,7 +1011,7 @@
 			if ((dev->class == ATA_DEV_ATAPI) &&
 			    (dev->flags & ATA_DFLAG_AN)) {
 				ata_scsi_media_change_notify(dev);
-				sata_fm_event_post(&ap->link, dev, FM_EVT_MEDIA_CHANGE);
+				sata_fm_event_post(&ap->link, dev,FM_SATAEH);
 			}	
 			return 0;
 		} else {
@@ -1035,7 +1035,7 @@
 			if ((link->device->class == ATA_DEV_ATAPI) &&
 			    (link->device->flags & ATA_DFLAG_AN)) {
 				ata_scsi_media_change_notify(link->device);
-				sata_fm_event_post(link, link->device, FM_EVT_MEDIA_CHANGE);
+				sata_fm_event_post(link, link->device, FM_SATAEH);
 			}
 		}
 
diff -Nur kernel-2.6.28.10-vs2.3.0.36.11.85b15/drivers/ata/libata-pmp.c kernel-2.6.28.10-vs2.3.0.36.11.85b15_fmd/drivers/ata/libata-pmp.c
--- kernel-2.6.28.10-vs2.3.0.36.11.85b15/drivers/ata/libata-pmp.c	2011-03-29 18:04:19.000000000 +0800
+++ kernel-2.6.28.10-vs2.3.0.36.11.85b15_fmd/drivers/ata/libata-pmp.c	2011-06-30 11:42:18.000000000 +0800
@@ -839,7 +839,7 @@
 //  faultname = ata_err_string(qc->err_mask);       /* qc? */
 	faultname = ata_err_string(link->eh_info.err_mask);
 
-	if (flag == FM_EVT_MEDIA_CHANGE)
+	if (flag == FM_SATAEH )
 		(void) snprintf(fullclass, sizeof(fullclass), "%s.io.sata.disk.media_change",
 			FM_EREPORT_CLASS);
 	else {
@@ -853,14 +853,14 @@
 	}
 
 	evt->flags = flag;
-	evt->device = fm_device;
-	evt->fullclass = fullclass;
+	strcpy(evt->devid, fm_device);
+	strcpy(evt->eclass, fullclass);
 
 	fm_event_post_intr(evt);
 	printk("sata_fm_event_post_intr(%s)\n", fullclass);
 
-	kfree(evt->device);
-	kfree(evt->fullclass);
+	kfree(fm_device);
+	kfree(fullclass);
 	kfree(evt);
 }
 
diff -Nur kernel-2.6.28.10-vs2.3.0.36.11.85b15/drivers/pci/pcie/aer/aerdrv_errprint.c kernel-2.6.28.10-vs2.3.0.36.11.85b15_fmd/drivers/pci/pcie/aer/aerdrv_errprint.c
--- kernel-2.6.28.10-vs2.3.0.36.11.85b15/drivers/pci/pcie/aer/aerdrv_errprint.c	2011-03-29 18:04:19.000000000 +0800
+++ kernel-2.6.28.10-vs2.3.0.36.11.85b15_fmd/drivers/pci/pcie/aer/aerdrv_errprint.c	2011-06-30 10:59:55.000000000 +0800
@@ -190,7 +190,7 @@
 	char * loglevel;
 	char *str;
 	fm_evt.type = FM_PCIEAER;
-	str = fm_evt.buf;
+	str = fm_evt.ev_data;
 	memset(str, 0, 100);
 	if (info->severity == AER_CORRECTABLE)
 	{
diff -Nur kernel-2.6.28.10-vs2.3.0.36.11.85b15/fm/fm_proc.c kernel-2.6.28.10-vs2.3.0.36.11.85b15_fmd/fm/fm_proc.c
--- kernel-2.6.28.10-vs2.3.0.36.11.85b15/fm/fm_proc.c	2011-03-29 18:04:19.000000000 +0800
+++ kernel-2.6.28.10-vs2.3.0.36.11.85b15_fmd/fm/fm_proc.c	2011-06-30 11:32:28.000000000 +0800
@@ -18,7 +18,7 @@
 fm_proc_uc(struct fm_kevt *evt)
 {
 	char *ev_class = NULL;
-	ev_class = evt->fullclass;
+	strcpy(ev_class, evt->eclass);
 
 	printk("KFMD: Fatal event.\n");
 	printk("KFMD: KEXEC is going on.\n");
@@ -34,7 +34,7 @@
 fm_proc_faultevt(struct fm_kevt *evt)
 {
 	char *ev_class = NULL;
-	ev_class = evt->fullclass;
+	strcpy(ev_class, evt->eclass);
 
 	if (strcmp(ev_class, "ereport.cpu.intel.quickpath.mem_page_ce") == 0) {
 		printk("FM: Page offline.\n");
@@ -47,7 +47,7 @@
 	char *ev_class = NULL;
 	u8 ev_type;
 
-	ev_class = evt->fullclass;
+	strcpy(ev_class, evt->eclass);
 	ev_type = evt->flags;
 
 /*for debug*/
diff -Nur kernel-2.6.28.10-vs2.3.0.36.11.85b15/include/linux/fm/fm.h kernel-2.6.28.10-vs2.3.0.36.11.85b15_fmd/include/linux/fm/fm.h
--- kernel-2.6.28.10-vs2.3.0.36.11.85b15/include/linux/fm/fm.h	2011-03-29 18:04:19.000000000 +0800
+++ kernel-2.6.28.10-vs2.3.0.36.11.85b15_fmd/include/linux/fm/fm.h	2011-06-30 11:26:28.000000000 +0800
@@ -16,17 +16,23 @@
  */
 #define FM_MAX_MSGSIZE        1000
 
-#define FM_BEAT					1
-#define FM_DEBUG				2
-#define FM_ON					3
-#define FM_OFF					4
-#define FM_EVT_UC				5
-#define FM_EVT_RC				6
-#define FM_EVT_MEDIA_CHANGE		7
-#define FM_NET					10
-#define FM_MPIO					11
-#define FM_PCIEAER				12
+#define FM_EVT_UC		0
+#define FM_EVT_RC		1
 
+#define FM_BEAT			0x01
+#define FM_DEBUG		0x02
+#define FM_ON			0x03
+#define FM_OFF			0x04
+//#define FM_EVT_MEDIA_CHANGE	0x07
+
+#define FM_TOPO			0x0a
+#define FM_MCA			0x0b
+#define FM_SATAEH		0x0c
+#define FM_NET			0x0d
+#define FM_MPIO			0x0e
+#define FM_PCIEAER		0x0f
+
+#if 0
 struct fm_kevt {
 	__u8 flags;			/* RC or UC ? */
 	__u8 cpu;			/* processor id */
@@ -36,6 +42,23 @@
 	char *fullclass;		/* event (ereport.class.*) */
 	void *ev_data;			/* data */
 };
+#endif
+
+struct fm_kevt {
+	__u8 flags;		/* RC or UC ? */
+	__u16 type;		/* event type */
+
+	__u8 cpuid;		/* processor id */
+	__u64 addr;		/* physical address */
+	__u8 devid[64];		/* pci id string:  0000:00:00.0 */
+
+	__u8 eclass[64];		/* event (ereport.class.*) */
+	__u8 ev_data[256];	/* data */
+
+//	char *device;		/* device id */
+//	char *fullclass;	/* event (ereport.class.*) */
+//	void *ev_data;		/* data */
+};
 
 extern void fm_event_post_proc(struct fm_kevt *ev);
 extern int fm_event_post_intr(struct fm_kevt *evt);
